### DISCUSSION TITLE
Cleanlab license and compatibility with AutoGluon (Apache v2)

### DISCUSSION THREAD

**Discussion by willsmithorg** [cleanlab/cleanlab#102] [Q&A] [answered]
Hi,  author(s) of cleanlab.

I am considering making an enhancement to Autogluon, to automatically include label cleaning using Cleanlab as a (likely optional) preprocessing step.

However, before I start I'm concerned about the licenses:
- cleanlab is Affero GPL v3  (i.e. a GPL-style license)
- Autogluon is Apache v2   (i.e. a [BSD-style license](https://en.wikipedia.org/wiki/BSD_licenses))

My current understanding is that software using both types are 'incompatible', so that using cleanlab within Autogluon would require that Autogluon itself become GPL-style.  I am clearly not able to do that - I am but a very minor contributor.      But I may be wrong on the licensing.  This passage in the Affero GPL license may be relevant:

end of section 5:

> A compilation of a covered work with other separate and independent
works, which are not by their nature extensions of the covered work,
and which are not combined with it such as to form a larger program,
in or on a volume of a storage or distribution medium, is called an
"aggregate" if the compilation and its resulting copyright are not
used to limit the access or legal rights of the compilation's users
beyond what the individual works permit.  Inclusion of a covered work
in an aggregate does not cause this License to apply to the other
parts of the aggregate.

I am unclear whether calling an unmodified version of Cleanlab, via its existing API, from the source of Autogluon, would constitute a 'larger program'.  Even if so, and the larger program is an aggregate, then the above says that cleanlab's license would not apply to the Autogluon parts of the aggregate.

I do understand the motivations of both BSD-style and GPL-style licensing (some years ago I worked in open source policy and have attend met and discussed licensing with Richard Stallman, Georg Greve and David Axmark) and have no particular leaning towards which is "more free".  

I also notice that [earlier versions of cleanlab](https://github.com/cleanlab/cleanlab/commits/master/LICENSE) were released under the GPL v3 (non-GPL) and MIT licenses, so presumably you've been thinking hard about licensing.

Do you have any suggestions what my next steps might be?

Thanks
William Smith

**Comment by jwmueller** [comment, marked as answer]
Hi @willsmithorg 
I'm also a co-author of autogluon and been very happy to see all your contributions there :)
Also excited to see your interest in cleanlab as I've recently joined this team.  We've just decided to build a data-centric ML company around the cleanlab library, so stay tuned as many exciting developments are planned!  However that means cleanlab needs to keep the AGPL license for the time being while we are figuring things out.

While I'd love to see cleanlab be part of autogluon's ML pipeline, our team cannot provide legal advice regarding the implications of incorporating AGPL code into that Apache-v2 repo, so you'd have to figure it out with the core autogluon devs.  That said, our company intends to utilize autogluon, as it is a great tool, and to continue contributing to the autogluon repo and collaborating with their devs. So we'd also love it if the autogluon team can benefit from cleanlab as well!
